File: retaRust/python_arch_reference/reta_architecture/table_state.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
from __future__ import annotations

"""Mutable table-state sections for Reta.

Stage 26 separates the table runtime's raw mutable state from the ``Tables``
class itself.  ``Tables`` remains the legacy-compatible global table section,
but the state it owns is now grouped into explicit sections:

* highest-row bounds
* generated-column metadata
* display flags and religion-number state
* display-row -> original-row mapping

This is intentionally small and conservative: old attributes continue to point
to the same mutable dict/list objects, so legacy code can still mutate them in
place while the architecture can inspect the state as named sections.
"""

from collections import OrderedDict
from dataclasses import dataclass, field
from typing import Any, Callable, MutableMapping

try:
    from orderedset import OrderedSet
except (ModuleNotFoundError, ImportError):  # pragma: no cover - minimal env fallback
    OrderedSet = set


def _highest_rows(hoechst_zeil: int | None) -> dict[int, int]:
    if hoechst_zeil is None:
        return {1024: 1024, 114: 163}
    return {1024: hoechst_zeil, 114: hoechst_zeil}


@dataclass
class GeneratedColumnSection:
    """Mutable metadata section for generated table columns."""

    parameters: MutableMapping[int, Any] = field(default_factory=OrderedDict)
    tags: MutableMapping[int, Any] = field(default_factory=OrderedDict)

    def snapshot(self) -> dict:
        return {
            "class": type(self).__name__,
            "parameters_len": len(self.parameters),
            "tags_len": len(self.tags),
            "parameters_type": type(self.parameters).__name__,
            "tags_type": type(self.tags).__name__,
        }


@dataclass
class TableDisplayState:
    """Mutable display/session flags shared by prepare, combi and output layers."""

    keine_ueberschriften: bool = False
    keine_leeren_inhalte: bool = False
    spalte_gestirn: bool = False
    religion_numbers: list = field(default_factory=list)

    def snapshot(self) -> dict:
        return {
            "class": type(self).__name__,
            "keine_ueberschriften": self.keine_ueberschriften,
            "keine_leeren_inhalte": self.keine_leeren_inhalte,
            "spalte_gestirn": self.spalte_gestirn,
            "religion_numbers_len": len(self.religion_numbers),
        }


@dataclass
class TableStateSections:
    """Named mutable sections owned by a ``Tables`` instance."""

    highest_rows: dict[int, int]
    display: TableDisplayState = field(default_factory=TableDisplayState)
    generated_columns: GeneratedColumnSection = field(default_factory=GeneratedColumnSection)
    row_display_to_original: MutableMapping[int, int] = field(default_factory=OrderedDict)
    generated_rows_factory: Callable[[], Any] = OrderedSet

    def new_generated_rows(self):
        return self.generated_rows_factory()

    def snapshot(self) -> dict:
        return {
            "class": type(self).__name__,
            "highest_rows": dict(self.highest_rows),
            "display": self.display.snapshot(),
            "generated_columns": self.generated_columns.snapshot(),
            "row_display_to_original_len": len(self.row_display_to_original),
            "generated_rows_factory": getattr(self.generated_rows_factory, "__name__", str(self.generated_rows_factory)),
        }


@dataclass(frozen=True)
class TableStateBundle:
    """Factory/metadata object for table-state sections."""

    ordered_dict_factory: Callable[[], MutableMapping] = OrderedDict
    ordered_set_factory: Callable[[], Any] = OrderedSet

    def create_sections(self, hoechst_zeil: int | None = None) -> TableStateSections:
        return TableStateSections(
            highest_rows=_highest_rows(hoechst_zeil),
            generated_columns=GeneratedColumnSection(
                parameters=self.ordered_dict_factory(),
                tags=self.ordered_dict_factory(),
            ),
            row_display_to_original=self.ordered_dict_factory(),
            generated_rows_factory=self.ordered_set_factory,
        )

    def snapshot(self) -> dict:
        return {
            "class": type(self).__name__,
            "sections": [
                "highest_rows",
                "display",
                "generated_columns",
                "row_display_to_original",
                "generated_rows",
            ],
            "architecture_owner": "reta_architecture.table_state",
            "legacy_owner": "reta_architecture.table_runtime.Tables",
        }


def bootstrap_table_state(
    ordered_dict_factory: Callable[[], MutableMapping] | None = None,
    ordered_set_factory: Callable[[], Any] | None = None,
) -> TableStateBundle:
    return TableStateBundle(
        ordered_dict_factory=ordered_dict_factory or OrderedDict,
        ordered_set_factory=ordered_set_factory or OrderedSet,
    )

File: retaRust/crates/reta_architecture/src/table_state.rs
//! Mutable table-state sections transcompiled from
//! `python_arch_reference/reta_architecture/table_state.py`.
//!
//! The Python architecture keeps these sections mutable because the historical
//! runtime mutates table state in-place.  The Rust version names the same
//! sections explicitly and gives later table/output ports a single owner for
//! highest-row bounds, generated-column metadata and display flags.

use std::collections::{BTreeMap, BTreeSet};

use serde::{Deserialize, Serialize};

pub fn highest_rows(highest_row: Option<i64>) -> BTreeMap<i64, i64> {
    match highest_row {
        Some(value) => BTreeMap::from([(1024, value), (114, value)]),
        None => BTreeMap::from([(1024, 1024), (114, 163)]),
    }
}

#[derive(Clone, Debug, Default, Eq, PartialEq, Serialize, Deserialize)]
pub struct GeneratedColumnSection {
    pub parameters: BTreeMap<i64, Vec<String>>,
    pub tags: BTreeMap<i64, Vec<String>>,
}

impl GeneratedColumnSection {
    pub fn snapshot(&self) -> GeneratedColumnSectionSnapshot {
        GeneratedColumnSectionSnapshot {
            class: "GeneratedColumnSection".to_string(),
            parameters_len: self.parameters.len(),
            tags_len: self.tags.len(),
            parameters_type: "BTreeMap".to_string(),
            tags_type: "BTreeMap".to_string(),
        }
    }
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct GeneratedColumnSectionSnapshot {
    pub class: String,
    pub parameters_len: usize,
    pub tags_len: usize,
    pub parameters_type: String,
    pub tags_type: String,
}

#[derive(Clone, Debug, Default, Eq, PartialEq, Serialize, Deserialize)]
pub struct TableDisplayState {
    pub keine_ueberschriften: bool,
    pub keine_leeren_inhalte: bool,
    pub spalte_gestirn: bool,
    pub religion_numbers: Vec<i64>,
}

impl TableDisplayState {
    pub fn snapshot(&self) -> TableDisplayStateSnapshot {
        TableDisplayStateSnapshot {
            class: "TableDisplayState".to_string(),
            keine_ueberschriften: self.keine_ueberschriften,
            keine_leeren_inhalte: self.keine_leeren_inhalte,
            spalte_gestirn: self.spalte_gestirn,
            religion_numbers_len: self.religion_numbers.len(),
        }
    }
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct TableDisplayStateSnapshot {
    pub class: String,
    pub keine_ueberschriften: bool,
    pub keine_leeren_inhalte: bool,
    pub spalte_gestirn: bool,
    pub religion_numbers_len: usize,
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct TableStateSections {
    pub highest_rows: BTreeMap<i64, i64>,
    pub display: TableDisplayState,
    pub generated_columns: GeneratedColumnSection,
    pub row_display_to_original: BTreeMap<i64, i64>,
    pub generated_rows: BTreeSet<i64>,
}

impl TableStateSections {
    pub fn new(highest_row: Option<i64>) -> Self {
        Self {
            highest_rows: highest_rows(highest_row),
            display: TableDisplayState::default(),
            generated_columns: GeneratedColumnSection::default(),
            row_display_to_original: BTreeMap::new(),
            generated_rows: BTreeSet::new(),
        }
    }

    pub fn snapshot(&self) -> TableStateSectionsSnapshot {
        TableStateSectionsSnapshot {
            class: "TableStateSections".to_string(),
            highest_rows: self.highest_rows.clone(),
            display: self.display.snapshot(),
            generated_columns: self.generated_columns.snapshot(),
            row_display_to_original_len: self.row_display_to_original.len(),
            generated_rows_len: self.generated_rows.len(),
            generated_rows_factory: "BTreeSet".to_string(),
        }
    }
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct TableStateSectionsSnapshot {
    pub class: String,
    pub highest_rows: BTreeMap<i64, i64>,
    pub display: TableDisplayStateSnapshot,
    pub generated_columns: GeneratedColumnSectionSnapshot,
    pub row_display_to_original_len: usize,
    pub generated_rows_len: usize,
    pub generated_rows_factory: String,
}

#[derive(Clone, Debug, Default, Eq, PartialEq, Serialize, Deserialize)]
pub struct TableStateBundle;

impl TableStateBundle {
    pub fn create_sections(&self, highest_row: Option<i64>) -> TableStateSections {
        TableStateSections::new(highest_row)
    }

    pub fn snapshot(&self) -> TableStateBundleSnapshot {
        TableStateBundleSnapshot {
            class: "TableStateBundle".to_string(),
            sections: vec![
                "highest_rows".to_string(),
                "display".to_string(),
                "generated_columns".to_string(),
                "row_display_to_original".to_string(),
                "generated_rows".to_string(),
            ],
            architecture_owner: "reta_architecture.table_state".to_string(),
            legacy_owner: "reta_architecture.table_runtime.Tables".to_string(),
        }
    }
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct TableStateBundleSnapshot {
    pub class: String,
    pub sections: Vec<String>,
    pub architecture_owner: String,
    pub legacy_owner: String,
}

pub fn bootstrap_table_state() -> TableStateBundle {
    TableStateBundle
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn default_highest_rows_match_python() {
        assert_eq!(highest_rows(None), BTreeMap::from([(1024, 1024), (114, 163)]));
        assert_eq!(highest_rows(Some(42)), BTreeMap::from([(1024, 42), (114, 42)]));
    }

    #[test]
    fn created_sections_share_named_state() {
        let sections = bootstrap_table_state().create_sections(None);
        let snapshot = sections.snapshot();
        assert_eq!(snapshot.highest_rows[&1024], 1024);
        assert_eq!(snapshot.generated_columns.parameters_len, 0);
    }
}


// Stage 15: explicit py-reta-arch compatibility surface markers.
// These markers keep historical Python architecture symbol names visible
// while the Rust implementation is migrated module by module. They are
// not a claim of byte-exact semantic replacement for every listed helper.
#[allow(dead_code)]
pub const PY_ARCH_STAGE15_SURFACE: &[&str] = &[
    "_highest_rows",
    "new_generated_rows",
];

#[allow(dead_code)]
pub fn stage15_py_surface_names() -> &'static [&'static str] {
    PY_ARCH_STAGE15_SURFACE
}
